doc/neps/nep-0050-scalar-promotion.rst

```diff
@@ -0,0 +1,743 @@
+:Created: 2021-01-10
+
+
+Abstract
+========
+
+Since NumPy 1.7, promotion rules are defined through the "safe casting"
+concept which relies on inspection of the actual values involved.
+While these choices were well intended, they lead to complexity: both
+in implementation and user experience.
+
+There are two kinds of confusing results:
+
+1. Value-based promotion means that values determine output types::
+
+     np.result_type(np.int8, 1) == np.int8
+     np.result_type(np.int8, 255) == np.int16
+
+   This also holds when working with 0-D arrays (so-called "scalar arrays")::
+
+     int64_0d_array = np.array(1, dtype=np.int64)
+     np.result_type(np.int8, int64_0d_array) == np.int8
+
+   This logic arises because ``1`` can be represented by an ``int8`` while
+   ``255`` can be represented by an ``int16`` *or* ``uint8``.
+
+   Because of this, the exact type is often ignored for 0-D arrays or
+   NumPy scalars.
+
+2. For a Python ``int``, ``float``, or ``complex`` the value is inspected as
+   previously.  But surprisingly *not* when the NumPy object is a 0-D array
+   or NumPy scalar::
+
+     np.result_type(np.array(1, dtype=np.uint8), 1) == np.int64
+     np.result_type(np.int8(1), 1) == np.int64
+
+   The reason is that the special value-based promotion is disabled when all
+   objects are scalars or 0-D arrays.
+   NumPy thus returns the same type as ``np.array(1)``, which is usually
+   an ``int64`` (this depends on the system).
+
+Note that the examples apply also to operations like multiplication,
+addition, comparisons, and their corresponding functions like ``np.multiply``.
+
+This NEP proposes to refactor the behaviour around two guiding principles:
+
+1. Values must never influence result type.
+2. NumPy scalars and 0-D arrays should behave consistently with their
+   N-D counterparts.
+
+We propose to remove all value-based logic and add special handling for
+Python scalars to preserve some convenient behaviors.
+Python scalars will be considered "weakly" typed.
+When a NumPy array/scalar is combined with a Python scalar, it will
+be converted to the NumPy dtype, such that::
+
+    np.array([1, 2, 3], dtype=np.uint8) + 1  # returns a uint8 array
+    np.array([1, 2, 3], dtype=np.float32) + 2.  # returns a float32 array
+
+There will be no dependence on the Python value itself.
+
+The proposed changes also apply to ``np.can_cast(100, np.int8)``, however,
+we expect that the behaviour in functions (promotion) will, in practice, be far
+more important than the casting change itself.
+
+
+Schema of the new proposed promotion rules
+------------------------------------------
+
+After the change, the promotions in NumPy will follow the schema below.
+Promotion always occurs along the green lines:
+from left to right within their kind and to a higher kind only when
+necessary.
+The result kind is always the largest kind of the inputs.
+Note that NumPy considers ``float32`` to have a lower precision than ``int32``
```

Comment:
The reason is that float32 cannot exactly represent all int32 values, so it makes sense.  The implication in the figure is that I moved float32 slightly to the left of int32.

It is not really safe to cast int32 to float32 or int64 to float64 and because of that promotion avoids int32 to float32 (it goes to float64).  But we make the exception that we consider int64 to float64 a safe cast and a valid promotion (even though it arguably loses precision).

Reply:
The implication on the figure is that the float32 bubble is to the left of the in32 to bubble